packages/styled-components/src/models/StyledComponent.ts

```diff
@@ -29,24 +31,40 @@ import isTag from '../utils/isTag';
 import joinStrings from '../utils/joinStrings';
 import merge from '../utils/mixinDeep';
 import ComponentStyle from './ComponentStyle';
+import Keyframes from './Keyframes';
 import { useStyleSheet, useStylis } from './StyleSheetManager';
 import { ThemeContext } from './ThemeProvider';
 
-const identifiers: { [key: string]: number } = {};
+/** Shared context for component identifiers, preventing duplicate ids */
+const COMPONENT_IDENTIFIERS: Record<string, Record<string, number>> = {}
+
 
 /* We depend on components having unique IDs */
-function generateId(displayName?: string, parentComponentId?: string): string {
-  const name = typeof displayName !== 'string' ? 'sc' : escape(displayName);
-  // Ensure that no displayName can lead to duplicate componentIds
-  identifiers[name] = (identifiers[name] || 0) + 1;
+function generateId<OuterProps extends object>(rules: RuleSet<OuterProps>, displayName?: string, parentComponentId?: string): string {
+  const namespace = typeof displayName !== 'string' ? 'sc' : escape(displayName);
+
+  /** Keyframes cannot be interpolated without stylesheet */
+  const isNotKeyframe = (interpolation: Interpolation<OuterProps>) => !(interpolation instanceof Keyframes)
+  const styleSheet = flatten(rules).filter(isNotKeyframe).join('')
+
+  // SC_VERSION gives us isolation between multiple runtimes on the page at once
+  // this is improved further with use of the babel plugin "namespace" feature
+  const componentId = `${namespace}-${generateComponentId(SC_VERSION + namespace + styleSheet)}`;
```

Comment:
I removed the `name` from here, since styles should already be enough for hashing... not sure if it should be put back. Or maybe the `namespace` could be removed too, since it's prefixed anyway.

Reply:
Actually, might be more robust to use `name` instead of `namespace` here.

Reply:
Oops, my own self-review confusion; don't need `namespace` in hash since it's prefixed anyway.

Reply:
Fixed!